Add user facing setting for persisting headers

TheMightyPenguin opened the pull request:
Resolves https://github.com/graphql/graphiql/issues/2793

# Description

Adds a user control in the settings dialog for persisting headers. This is only visible if the `shouldPersistHeaders` is explicitly set as `false`.

Demo and brief explanation: https://www.loom.com/share/d72cf65025b04b1f9d6d5e518445bc25

<img width="716" alt="image" src="https://user-images.githubusercontent.com/11748696/202808224-b5410c29-6964-4664-88ac-e34a34b30f8d.png">


I also found 2 potential issues when working on this:
- https://github.com/graphql/graphiql/issues/2891 (Accessibility related, I think the ButtonGroup we use in the settings dialog doesn't have the correct accessibility behavior and attributes. I decided to keep it as a separate issue to keep this PR smaller, I'll work on that one once this one is completed)
- https://github.com/graphql/graphiql/issues/2892

**Note for the future:** I think it would be nice to somehow refactor this out into a user settings module, not the UI but the logic to store and access user settings. At the moment we have Theme and now Persist Headers, but there could be more user controlled settings going forward and having a central place for this logic (including defaults and such) could be nice.

### Questions
1. As headers are not persisted as a security measure, should we include a warning for the user in the dialog?
2. I added logic to cleanup the headers storage if this option is turned off, do we also need to do the same for tabs state?

acao commented:
thank you so much for this @TheMightyPenguin !

acao commented:
> should we include a warning for the user in the dialog?
I think so, perhaps just a warning to make sure they trust the device they are persisting these secrets to.

> do we also need to do the same for tabs state?
it would be lovely if you could cover that case as well!

acao commented:
@TheMightyPenguin merged an upstream PR where @jonathanawesome fixed the `act()` issue, sorry for the merge conflicts!

acao commented:
> This is only visible if the `shouldPersistHeaders` is not passed to the GraphQL component.

This seems like a sensible choice, however - what if this prop just determines the default value, and the user can choose to override it?

TheMightyPenguin commented:
> > This is only visible if the `shouldPersistHeaders` is not passed to the GraphQL component.
> 
> this seems sensible! what do you think of this?
> 
> * if `shouldPersistHeaders` is`false` explicitly, we should not show the setting
> * otherwise, if it is `true` or not specified, the user defined setting is available, defaulting to enabled/disabled respectively

I like it, seems considerate with the end user being in control of their privacy, I'll include this logic in the next commit!

TheMightyPenguin commented:
@acao updated PR, now we only show the setting if the property is not set (`undefined`) or `true`, and when it's set to false we also clean up the tabs storage, removing any headers key in there, I added tests for this logic.

I also added a warning message (using the warning color, let me know is the styling is not right 😅):
<img width="716" alt="image" src="https://user-images.githubusercontent.com/11748696/202808277-b4429ba2-64e8-464f-8865-b3d086c16bc1.png">

thomasheyenbrock reviewed (changes requested):
I'm not sure if we should add the state for user-controlling this setting to the `EditorContextProvider`. I feel that this is more of a thing that should happen in user-land (from the perspective of `@graphiql/react`).

To give an example: If I'm building my own custom GraphQL IDE with the components from `@graphiql/react`, I might have a clear opinion on how I want to persist headers or not. I already have the `shouldPersistHeaders` prop that I can use to control this behavior. Changing this state with the `setShouldPersistHeaders` method added in this PR gives me too much control that I don't need.

My proposal here is to move the state for the new toggle button in the settings dialog to the `GraphiQL` component. This component can combine this state with the value it receives as `shouldPersistHeaders` prop and pass the "final" value for that prop to the `GraphiQLProvider`. It also passes the value and the state-setter down to the `GraphiQLInterface` component which will use it to render the toggle button in the settings dialog. That way we don't need to expose any more state or methods from the `EditorContext` which keep it more easy to use.

Finally: Having a chunk of logic in the `EditorContextProvider` that clears the persisted header state when the `shouldPersistHeaders` prop changes is a fantastic addition, so we should definitely leave that in! (In fact, I would even consider it a bug that we're not doing it yet.)

TheMightyPenguin commented:
> I'm not sure if we should add the state for user-controlling this setting to the `EditorContextProvider`. I feel that this is more of a thing that should happen in user-land (from the perspective of `@graphiql/react`).
> 
> To give an example: If I'm building my own custom GraphQL IDE with the components from `@graphiql/react`, I might have a clear opinion on how I want to persist headers or not. I already have the `shouldPersistHeaders` prop that I can use to control this behavior. Changing this state with the `setShouldPersistHeaders` method added in this PR gives me too much control that I don't need.
> 
> My proposal here is to move the state for the new toggle button in the settings dialog to the `GraphiQL` component. This component can combine this state with the value it receives as `shouldPersistHeaders` prop and pass the "final" value for that prop to the `GraphiQLProvider`. It also passes the value and the state-setter down to the `GraphiQLInterface` component which will use it to render the toggle button in the settings dialog. That way we don't need to expose any more state or methods from the `EditorContext` which keep it more easy to use.
> 
> Finally: Having a chunk of logic in the `EditorContextProvider` that clears the persisted header state when the `shouldPersistHeaders` prop changes is a fantastic addition, so we should definitely leave that in! (In fact, I would even consider it a bug that we're not doing it yet.)

@thomasheyenbrock I tried implementing this, but there's an issue as the `GraphiQL` would not have access to the storage context. We could make this work somehow but not sure how clean the solution would be 😅. Gonna think this one a bit more to see what workaround we could use.

Additionally, I'd say we could always show the setting in the GraphiQL component, removing the logic we have for showing this in only certain scenarios (e.g. when the prop is `true` or `undefined`) would simplify the code.

TheMightyPenguin commented:
@acao @thomasheyenbrock pushed a few commits addressing the comments!

acao reviewed (approved):
Everything looks in order to me, and I concur with your choice to make the option always available!

jonathanawesome commented:
Looks good to me, thanks @TheMightyPenguin for the thorough PR!

TheMightyPenguin commented:
@acao @thomasheyenbrock @jonathanawesome anything missing from my side in this PR? (just making sure I am not blocking anything)

thomasheyenbrock reviewed (approved):
Awesome, this looks fine for me 💯
